File: wagerproof-ios-native/WagerproofKit/Sources/WagerproofStores/MLBSeriesSignalsStore.swift
import Foundation
import Observation
import Supabase
import WagerproofModels
import WagerproofServices

/// MLB series-position signals (G2/G3 carryover) from `mlb_game_signals`.
/// Mirrors RN `hooks/useMLBSeriesSignals.ts`: each `home_signals` /
/// `away_signals` entry is a JSON-encoded string; only entries whose parsed
/// `category == "series"` (with a message) survive.
@Observable
@MainActor
public final class MLBSeriesSignalsStore {
    public private(set) var signals: [MLBSeriesSignal] = []
    public private(set) var loading: Bool = false
    public private(set) var errorMessage: String?
    public private(set) var lastFetched: Date?

    /// 5-minute stale window — matches RN `staleTime: 5 * 60 * 1000`.
    private let staleWindow: TimeInterval = 5 * 60

    public init() {}

    private struct GameSignalsRow: Decodable {
        let gamePk: Int
        let homeTeamName: String
        let awayTeamName: String
        let homeSignals: [String]?
        let awaySignals: [String]?

        enum CodingKeys: String, CodingKey {
            case gamePk = "game_pk"
            case homeTeamName = "home_team_name"
            case awayTeamName = "away_team_name"
            case homeSignals = "home_signals"
            case awaySignals = "away_signals"
        }
    }

    public func refreshIfStale(force: Bool = false) async {
        if !force, let last = lastFetched, Date().timeIntervalSince(last) < staleWindow {
            return
        }
        await refresh()
    }

    public func refresh() async {
        loading = true
        errorMessage = nil
        do {
            let cfb = await CFBSupabase.shared.client
            let rows: [GameSignalsRow] = try await cfb
                .from("mlb_game_signals")
                .select("game_pk, home_team_name, away_team_name, home_signals, away_signals")
                .execute()
                .value
            self.signals = MLBSeriesSignalsStore.parse(rows: rows)
            self.lastFetched = Date()
        } catch {
            errorMessage = "Failed to load series signals."
        }
        loading = false
    }

    private static func parse(rows: [GameSignalsRow]) -> [MLBSeriesSignal] {
        var out: [MLBSeriesSignal] = []
        for row in rows {
            let matchup = "\(row.awayTeamName) @ \(row.homeTeamName)"
            func collect(_ raws: [String]?, side: String, team: String) {
                for raw in raws ?? [] {
                    guard let data = raw.data(using: .utf8),
                          let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                          obj["category"] as? String == "series",
                          let message = obj["message"] as? String, !message.isEmpty
                    else { continue } // skip malformed entries silently, RN parity
                    let severity = (obj["severity"] as? String) == "positive" ? "positive" : "negative"
                    out.append(MLBSeriesSignal(
                        gamePk: row.gamePk,
                        matchup: matchup,
                        teamName: team,
                        teamSide: side,
                        severity: severity,
                        message: message
                    ))
                }
            }
            collect(row.homeSignals, side: "home", team: row.homeTeamName)
            collect(row.awaySignals, side: "away", team: row.awayTeamName)
        }
        return out
    }

    #if DEBUG
    public func debugSet(signals: [MLBSeriesSignal]) {
        self.signals = signals
        self.lastFetched = Date()
    }
    #endif
}

File: wagerproof-android-native/core/stores/src/main/kotlin/com/wagerproof/core/stores/MLBSeriesSignalsStore.kt
package com.wagerproof.core.stores

import androidx.compose.runtime.Stable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.wagerproof.core.models.MLBSeriesSignal
import com.wagerproof.core.models.serialization.FlexibleIntOrZeroSerializer
import com.wagerproof.core.services.BuildFlags
import com.wagerproof.core.services.SupabaseClients
import io.github.jan.supabase.postgrest.from
import io.github.jan.supabase.postgrest.query.Columns
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

/**
 * MLB series-position signals (G2/G3 carryover) from `mlb_game_signals`. Port
 * of iOS `MLBSeriesSignalsStore.swift` (mirrors RN `hooks/useMLBSeriesSignals.ts`):
 * each `home_signals` / `away_signals` entry is a JSON-encoded string; only
 * entries whose parsed `category == "series"` (with a message) survive.
 * 5-minute stale window.
 *
 * No coroutine scope — `refresh()` is a suspend fn called from the view.
 */
@Stable
class MLBSeriesSignalsStore {
    var signals by mutableStateOf<List<MLBSeriesSignal>>(emptyList()); private set
    var loading by mutableStateOf(false); private set
    var errorMessage by mutableStateOf<String?>(null); private set
    var lastFetched by mutableStateOf<Long?>(null); private set

    /** 5-minute stale window — matches RN `staleTime: 5 * 60 * 1000`. */
    private val staleWindowMs: Long = 5 * 60 * 1000

    @Serializable
    private data class GameSignalsRow(
        @Serializable(with = FlexibleIntOrZeroSerializer::class)
        @SerialName("game_pk") val gamePk: Int = 0,
        @SerialName("home_team_name") val homeTeamName: String = "",
        @SerialName("away_team_name") val awayTeamName: String = "",
        @SerialName("home_signals") val homeSignals: List<String>? = null,
        @SerialName("away_signals") val awaySignals: List<String>? = null,
    )

    suspend fun refreshIfStale(force: Boolean = false) {
        val last = lastFetched
        if (!force && last != null && System.currentTimeMillis() - last < staleWindowMs) return
        refresh()
    }

    suspend fun refresh() {
        loading = true
        errorMessage = null
        try {
            val cfb = SupabaseClients.cfb
            val rows: List<GameSignalsRow> = cfb
                .from("mlb_game_signals")
                .select(
                    columns = Columns.raw(
                        "game_pk, home_team_name, away_team_name, home_signals, away_signals",
                    ),
                )
                .decodeList()
            signals = parse(rows)
            lastFetched = System.currentTimeMillis()
        } catch (e: Exception) {
            errorMessage = "Failed to load series signals."
        }
        loading = false
    }

    fun debugSet(signals: List<MLBSeriesSignal>) {
        if (!BuildFlags.isDebugBuild) return
        this.signals = signals
        this.lastFetched = System.currentTimeMillis()
    }

    private companion object {
        private val json = Json { ignoreUnknownKeys = true; isLenient = true }

        fun parse(rows: List<GameSignalsRow>): List<MLBSeriesSignal> {
            val out = mutableListOf<MLBSeriesSignal>()
            for (row in rows) {
                val matchup = "${row.awayTeamName} @ ${row.homeTeamName}"

                fun collect(raws: List<String>?, side: String, team: String) {
                    for (raw in raws ?: emptyList()) {
                        // Skip malformed entries silently (RN parity).
                        val obj = runCatching { json.parseToJsonElement(raw).jsonObject }.getOrNull() ?: continue
                        val category = obj["category"]?.jsonPrimitive?.contentOrNull
                        if (category != "series") continue
                        val message = obj["message"]?.jsonPrimitive?.contentOrNull
                        if (message.isNullOrEmpty()) continue
                        val severity =
                            if (obj["severity"]?.jsonPrimitive?.contentOrNull == "positive") "positive" else "negative"
                        out.add(
                            MLBSeriesSignal(
                                gamePk = row.gamePk,
                                matchup = matchup,
                                teamName = team,
                                teamSide = side,
                                severity = severity,
                                message = message,
                            ),
                        )
                    }
                }

                collect(row.homeSignals, side = "home", team = row.homeTeamName)
                collect(row.awaySignals, side = "away", team = row.awayTeamName)
            }
            return out
        }
    }
}
